File: apps/yryvu-app/src/ipc/sparse_checkout.ts
// SPDX-License-Identifier: AGPL-3.0-or-later

import { invoke } from "@tauri-apps/api/core";

/// Mirrors `yryvu_bridge::repo::sparse_checkout::SparseCheckoutState`.
/// `patterns` are the non-blank / non-comment lines from
/// `.git/info/sparse-checkout`, order preserved.
export interface SparseCheckoutState {
  enabled: boolean;
  coneMode: boolean;
  patterns: string[];
}

/// Read the repo's current sparse-checkout state. All fields false /
/// empty when sparse is not enabled.
export function getSparseCheckoutState(
  repoPath: string,
): Promise<SparseCheckoutState> {
  return invoke<SparseCheckoutState>("get_sparse_checkout_state", { repoPath });
}

/// `git sparse-checkout init [--cone]` — enables sparse-checkout for
/// the repo and seeds the patterns file. Does not write user patterns
/// until `sparseSetPatterns` follows.
export function sparseInit(
  repoPath: string,
  coneMode: boolean,
): Promise<void> {
  return invoke<void>("sparse_init", { repoPath, coneMode });
}

/// `git sparse-checkout set [--cone] <patterns...>` — overwrites the
/// patterns file and updates the working tree to match.
export function sparseSetPatterns(
  repoPath: string,
  coneMode: boolean,
  patterns: string[],
): Promise<void> {
  return invoke<void>("sparse_set_patterns", {
    repoPath,
    coneMode,
    patterns,
  });
}

/// `git sparse-checkout disable` — restores every file and turns the
/// `core.sparseCheckout` bit off.
export function sparseDisable(repoPath: string): Promise<void> {
  return invoke<void>("sparse_disable", { repoPath });
}

/// `git sparse-checkout reapply` — re-runs the existing patterns
/// against the working tree (useful after pull / fetch).
export function sparseReapply(repoPath: string): Promise<void> {
  return invoke<void>("sparse_reapply", { repoPath });
}

File: crates/yryvu-bridge/src/commands/sparse_checkout.rs
// SPDX-License-Identifier: AGPL-3.0-or-later

//! Tauri commands for sparse checkout (issue #309).

use std::path::PathBuf;

use crate::repo::sparse_checkout::{
    apply_disable as apply_disable_impl, apply_init as apply_init_impl,
    apply_reapply as apply_reapply_impl, apply_set as apply_set_impl, read_state,
    SparseCheckoutState,
};

#[tauri::command]
pub async fn get_sparse_checkout_state(repo_path: String) -> Result<SparseCheckoutState, String> {
    tauri::async_runtime::spawn_blocking(move || read_state(&PathBuf::from(repo_path)))
        .await
        .map_err(|e| e.to_string())?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn sparse_init(repo_path: String, cone_mode: bool) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        apply_init_impl(&PathBuf::from(repo_path), cone_mode)
    })
    .await
    .map_err(|e| e.to_string())?
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn sparse_set_patterns(
    repo_path: String,
    cone_mode: bool,
    patterns: Vec<String>,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        apply_set_impl(&PathBuf::from(repo_path), cone_mode, &patterns)
    })
    .await
    .map_err(|e| e.to_string())?
    .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn sparse_disable(repo_path: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || apply_disable_impl(&PathBuf::from(repo_path)))
        .await
        .map_err(|e| e.to_string())?
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn sparse_reapply(repo_path: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || apply_reapply_impl(&PathBuf::from(repo_path)))
        .await
        .map_err(|e| e.to_string())?
        .map_err(|e| e.to_string())
}
